pkg/inject/inject.go

```diff
@@ -624,6 +624,9 @@ func (conf *ResourceConfig) injectPodSpec(values *podPatch) {
 				values.Proxy.Capabilities.Add = append(values.Proxy.Capabilities.Add, string(add))
 			}
 			for _, drop := range sc.Capabilities.Drop {
+				if string(drop) == "NET_RAW" || string(drop) == "NET_ADMIN" {
```

Comment:
It looks like we first copy the main container's capabilities into the proxy capabilities (to ensure psp compliance) and then later copy those to the proxy-init container's capabilities (see line 669).  I wonder if it would be more correct to keep the proxy's capabilities as they are but remove the dropping of the NET_RAW and NET_ADMIN capabilities from the proxy-init container.  As far as I know, only the proxy-init container needs these capabilities.

Reply:
Tried moving the logic to where we populate the init container capabilities, but that didn't work. Upon looking closer, it looked like the sidecar injection never really happened probably because it was rejected by the `ValidatingAdmissionWebhook`. I might be wrong here but I think the psp `linkerd-linkerd-control-plane` which specifies the need for `NET_RAW` and `NET_ADMIN` capabilities validates both the init and proxy containers which is leading to this. Please let me know if I'm making sense here ✌️

Reply:
It's true that the PSP policy validates the `securityContext` for all the containers in a pod. But the `linkerd-linkerd-control-plane` PSP only applies to pods in the `linkerd` namespace, and what we're trying to address here are user pods (not linkerd's) that attempt to remove these NET_* capabilities. When testing a deployment, if you look at its ReplicaSet description you might see an error about it not being admitted because of a forbidden PSP. Usually this just means the ServiceAccount associated to that pod hasn't been assigned a PSP. Let me know if you'd like me to provide you with PSP resources and an example to make this work :wink:

Reply:
Hey @aryan9600 I'm not totally sure what issue you were running in to but I got it working with the following diff:

```
diff --git a/pkg/inject/inject.go b/pkg/inject/inject.go
index f6767cf65..3a7b3c579 100644
--- a/pkg/inject/inject.go
+++ b/pkg/inject/inject.go
@@ -624,9 +624,6 @@ func (conf *ResourceConfig) injectPodSpec(values *podPatch) {
                                values.Proxy.Capabilities.Add = append(values.Proxy.Capabilities.Add, string(add))
                        }
                        for _, drop := range sc.Capabilities.Drop {
-                               if string(drop) == "NET_RAW" || string(drop) == "NET_ADMIN" {
-                                       continue
-                               }
                                values.Proxy.Capabilities.Drop = append(values.Proxy.Capabilities.Drop, string(drop))
                        }
                }
@@ -666,7 +663,19 @@ func (conf *ResourceConfig) injectPodSpec(values *podPatch) {
 func (conf *ResourceConfig) injectProxyInit(values *podPatch) {

        // Fill common fields from Proxy into ProxyInit
-       values.ProxyInit.Capabilities = values.Proxy.Capabilities
+       if values.Proxy.Capabilities != nil {
+               values.ProxyInit.Capabilities = &l5dcharts.Capabilities{}
+               values.ProxyInit.Capabilities.Add = values.Proxy.Capabilities.Add
+               values.ProxyInit.Capabilities.Drop = []string{}
+               // The proxy init container needs NET_RAW and NET_ADMIN, so skip those
+               // if they are dropped by the main container.
+               for _, drop := range values.Proxy.Capabilities.Drop {
+                       if string(drop) == "NET_RAW" || string(drop) == "NET_ADMIN" {
+                               continue
+                       }
+                       values.ProxyInit.Capabilities.Drop = append(values.ProxyInit.Capabilities.Drop, string(drop))
+               }
+       }
        values.ProxyInit.SAMountPath = values.Proxy.SAMountPath

        if v := conf.pod.meta.Annotations[k8s.CloseWaitTimeoutAnnotation]; v != "" {
```

Reply:
@alpeb thanks for the explanation, that makes sense :D 
@kleimkuhler sorry, turns out i was just being stupid and making a silly error